examples/iterator-implementation.jl

```diff
@@ -0,0 +1,173 @@
+type EulerIntegrator <: AbstractIntegrator
+    # nothing in here yet
+end
+
+type EulerState
+    t
+    y
+    dy
+end
+
+output(state::EulerState) = state.t, state.y, state.dy
+
+# see we don't need a fancy constructor
+function solver(ode::ExplicitODE,
+                ::Type{EulerIntegrator};
+                options...)
+    return Problem(ode,EulerIntegrator())
+end
+
+function init(ode::ExplicitODE, integ::EulerIntegrator)
+    t0, y0 = ode.t0, ode.y0
+    dy0 = similar(ode.dy0)
+    ode.F!(t0,y0,dy0)           # fill in the values of the derivative
+    EulerState(t0,y0,dy0)
+end
+
+function onestep!(ode::ExplicitODE, integ::EulerIntegrator, state::EulerState)
+    t, y, dy = output(state)
+    dt = 0.1
+    y += dt*dy
+    t += dt
+    ode.F!(t0,y0,dy0)           # update the derivative
+    return cont
+end
+
+
+"""
+
+There are several problems with the above implementation.  First of
+all, it has a constant prescribed step size.  This could easily be
+fixed by changing the type definition and the `solver` to
+
+"""
+
+type EulerIntegrator <: AbstractIntegrator
+    initstep
+end
+
+function solver(ode::ExplicitODE,
+                ::Type{EulerIntegrator};
+                initstep = 0.1,
+                options...)
+    return Problem(ode,EulerIntegrator(initstep))
+end
+
+"""
+
+we should also change the line `dt = 0.1` in the `onestep!` function
+to `dt = stepper.initstep`.  Now we can run our integrator with a
+custom step size!
+
+"""
+
+sol = solver(ode,EulerIntegrator,initstep = 0.01)
+for (t,y) in sol
+    if t > 1
+        print(t,y)
+        break
+    end
+end
+
+"""
+
+Another issue is type stability, to make `EulerIntegrator` perform
+better we should explicitly annotate the fields in both
+`EulerIntegrator` and `EulerState` like this
+
+"""
+
+type EulerIntegrator{T,Y} <: AbstractIntegrator{T,Y}
+    initstep::T
+end
+
+type EulerState{T,Y}
+    t::T
+    y::Y
+    dy::Y
+end
+
+function solver(ode::ExplicitODE{T,Y},
+                ::Type{EulerIntegrator};
+                initstep::T = T(0.1),
+                options...)
+    return Problem(ode,EulerIntegrator{T,Y}(initstep))
+end
+
+
+"""
+
+But the `EulerState{T,Y}` is exactly the same as `Step` from
+`base.jl`, so we can simplify it a bit more
+
+"""
+
+type EulerState{T,Y}
+    step::Step{T,Y}
+end
+
+"""
+
+Once we do that, in the `init` we should change
+`EulerState(t0,y0,dy0)` to `EulerState(Step(t0,y0,dy0))` and redefine
+`output` to `output(state::EulerState)=output(state.step)`
```

Comment:
There is this fallback too https://github.com/JuliaODE/ODE.jl/blob/8fda07ce93b7848acc8df57ec84df645a89a4e42/src/base.jl#L121